File: python/longestNonRepSubstr.py
class Solution:
    # @param {string} s
    # @return {integer}
    def lengthOfLongestSubstring(self, s):
        if not s: return 0
        
        n = len(s)
        start, end = 0, 0
        maxLen = 1
        showed = {}
        
        while end < n:
            
            if s[end] not in showed or showed[s[end]] < start:
                showed[s[end]] = end
                print end, start
                maxLen = max(end - start + 1, maxLen)
            else:
                start = showed[s[end]] + 1
                showed[s[end]] = end
            end += 1

        return maxLen

s = Solution()
print s.lengthOfLongestSubstring("eee")

File: cpp/longestNonRepSubstr.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        std::vector<int> appearances;
        appearances.resize(256, -1);
        int left = 0, right = 0, max_len = 0;
        
        while (right < s.length()) {
            cout << appearances[0]<<endl;
            if (appearances[s[right]] >= left) {
                cout << appearances[s[right]]<<endl;
                left = appearances[s[right]] + 1;
            }
            
            appearances[s[right]] = right;
            max_len = max(max_len, right-left+1);
            right ++;
        }
        
        return max_len;
    }
};

int main(int argc, char **argv)
{
  map<char, int> appearances;
  string s = "c";
  Solution sol;
  cout << sol.lengthOfLongestSubstring(s) << endl;
  return 0;
}